superchain/superchain.go

```diff
@@ -59,6 +66,40 @@ type ChainConfig struct {
 	// Chain is a simple string to identify the chain, within its superchain context.
 	// This matches the resource filename, it is not encoded in the config file itself.
 	Chain string `yaml:"-"`
+
+	// Hardfork Configuration Overrides
+	HardForkConfiguration `yaml:",inline"`
+}
+
+// replaceMissingOverridesWithDefaults overwrites each unspecified hardfork activation time override
+// with the superchain default.
+func (c *ChainConfig) replaceMissingOverridesWithDefaults(s Superchain) {
+	dVal := reflect.ValueOf(c)
+	sVal := reflect.ValueOf(&s.Config.hardForkDefaults)
+
+	hfcVal := reflect.ValueOf(HardForkConfiguration{})
+	for i := 0; i < hfcVal.NumField(); i++ {
+		hardForkName := hfcVal.Type().Field(i).Name
+		overrideValue := dVal.Elem().FieldByName(hardForkName)
+
+		if overrideValue.Interface().(*uint64) == nil {
+			defaultValue := sVal.Elem().FieldByName(hardForkName)
+			overrideValue.Set(defaultValue)
+		}
+	}
+
+	// This achieves:
+	//
+	// if c.RegolithTime == nil {
+	// 	c.RegolithTime = s.Config.hardForkDefaults.RegolithTime
+	// }
+
+	// if c.CanyonTime == nil {
+	// 	c.CanyonTime = s.Config.hardForkDefaults.CanyonTime
+	// }
+	//
+	// ...etc for each field in HardForkConfiguration
+
```

Comment:
A simpler alternative to this would be to _not_ have any superchain-wide defaults, and instead _require_ each chain to specify every hard fork timestamp.

Reply:
If we want to continue this approach (using reflection) I am sure I can tidy this up even more / improve readability. The variable names aren't great and some of the reflection invocations could be abstracted a bit.

Reply:
Is there a way to always hydrate with the defaults & then set the overrides? Would that be cleaner than having to do this reflection.

Reply:
Conclusion: yes!

[set defaults first, then override by unmarshalling into nonempty struct](https://github.com/ethereum-optimism/superchain-registry/pull/101/commits/60086b977718e3cb6df27f83f55362c6dd287e32)

Reply:
We need a test to show this works as expected. I suspect that the defaults are being overwritten with `nil` if the `ChainConfig` does not specify an override.

Reply:
Tests have now been added and the implementation has been reverted back to this reflection approach.